Make the BC3 shaders compatible with GLSL ES version 310

nanley opened the pull request:
This is based on a collection of changes I had to make when I tried to compile these shaders in gles31 apps. Please let me know what you think.

nchery-intel committed:
Output to an rgba16ui image for the BC1 and BC4 shaders

Increases compatibility with GLSL ES version 310.

The ARB_shader_image_load_store extension says,

	For textures allocated by the GL, an image unit format is
	compatible with a texture internal format if they match by size.

The textures were allocated by the GL implementation, so this format
substitution is safe to do.

nchery-intel committed:
Conditionally disable the const keyword in the BC3 shaders

Increases compatibility with GLSL ES version 310.

nchery-intel committed:
Avoid implicit data type conversions in the BC3 shaders

Increases compatibility with GLSL ES version 310.

nchery-intel committed:
Set the version of the BC3 shaders to 310 es

The shaders are compatible with GLSL ES version 310. Require the
GL_ARB_ES3_1_compatibility extension and change the version directives
accordingly so that we get validation from the GL implementation.

nchery-intel committed:
Add precision qualifiers to the BC3 shaders

Increases compatibility with GLSL ES version 310.

nchery-intel committed:
Add PFG_RGBA16_UINT to the PixelFormat enum

To be used in the following commit.

nchery-intel committed:
Use constructors for array initialization in the BC1 shader

Increases compatibility with GLSL ES version 310.

darksylinc commented:
1. Could you do `#define const` instead? I prefer if const stays, rather than being removed. It's a very useful keyword that improves readability and maintenance
2. @reduz thoughts? Originally betsy was written to maximize Vulkan compatibility so that it can be ported to Godot (that mostly reduces to declaring `#version 450` on top though), but goals have changed a little. AFAIK this PR needs GLES 3.1 (not 3.0) due to the need for Compute Shader support.

nanley commented:
Updated `const`.

WRT to Vulkan compatibility, we shouldn't need to change the version directive right? AFAICT, `GL_KHR_vulkan_glsl` is [compatible](https://github.com/KhronosGroup/GLSL/blob/3a2c90a75b6950b8c725d8a3cfacba796f605013/extensions/khr/GL_KHR_vulkan_glsl.txt#L48) with `#version 310 es`. 

One downside I've noticed with the PR's usage of rg32 and rgba16 (https://github.com/darksylinc/betsy/pull/12/commits/c2f0e029a526527c60a7bdf6bffe98652f37f5e3)  is that it'd require a Vulkan app to set `VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT` during image creation. This could have a negative performance impact with some Vulkan drivers. One solution is to update the stitch shader to use rgba16. However, that would impact a number of other shaders outside of the BC3 scope.

nchery-intel committed:
Make the params uniform in the BC4 shader uint2

This is a better fit for indexing and for using as a boolean.

While we're here, avoid implicit data type conversions with the elements
in the uniform. This increases compatibility with GLSL ES version 310.
